ecosystem/sep-0018.md

```diff
@@ -50,6 +50,20 @@ wallet.eth
 * **Term REGEXP:** `[_a-z][_a-z0-9]*`
 * **Key REGEXP:** `[_a-z][_a-z0-9]*(\.[_a-z][_a-z0-9]*)*`
 
+## Namespace Registry
+
+To guarantee interoperability and avoid naming conflicts, namespaces must be
+used as defined in SEP documents.
+
+- `app`: Space dedicated to vendor-specific keys. For example, "example.org" can
+  define its keys under `app.example_org`.
+- `config`: Space dedicated to ecosystem-wide configuration. Those are values
+  that impact software logic.
+  - `config.memo_required`: [SEP29](sep-0029.md)
+  - `config.multisig`: [SEP19](sep-0019.md)
+- `profile`: Space dedicated to account identification.
+- `wallet`: Space dedicated to external ledger addresses.
```

Comment:
It's not clear to me what fields should be within `profile` or `wallet`.

Could you share an example of what fields they'll be?

I'm guessing they'll be fields shared by applications so to prevent naming conflicts and duplication of data in those shared fields the fields in these two namespaces will need to be specified too. I think we'd benefit from having a wallet or service chime in on what fields they'd want in these namespaces and if they'd use them.

Reply:
The idea is that fields in those namespaces need to be part of a SEP before we use them. (as for any namespace other than `app`)

`Profile` would definitely require its own SEP. It could be a place where to put the usual profile things such as an icon & a pseudo to improve inter-operability/harmonization between Stellar-based services. For more serious things, it could be used for GPG keys, KYC or for proofs such as the ones Keybase uses for domains/accounts/...
Examples: `profile.icon`, `profile.gpg`, `profile.domain`, ...

`Wallet` is to link external wallets from other chains. This is meant to be used by anchors as well as social/tipping applications.
Examples: `wallet.btc`, `wallet.eth`, `wallet.xrp`...

> I think we'd benefit from having a wallet or service chime in on what fields they'd want in these namespaces and if they'd use them.

Definitely

Reply:
cc @tyvdh

Reply:
@MisterTicot I'm still interested on input from @tyvdh on this, but given that we agree it would be worthwhile to get input about `profile` and `wallet` from wallets, if we agree on adding the app namespace I think we should move profile and wallet to a separate PR and ideally have some input from others for that. What do you think?

Reply:
I'm definitely on the side of not adding things which aren't being asked for or used in other forms elsewhere. I'm also ultimately not convinced adding lots of data entries is a better route than using an external reference like the TOML file. A standard format seems like a better SEP than a standard storage location. Storing lots of data attributes will get expensive for user's or wallets and with the fine grained options you're talking about, it seems to me like storing these options outside the ledger might be a better option.

Reply:
Thanks @tyvdh, that's a great point. Do you think this applies to the `app` namespace too or just the `profile` and `wallet` ones?

Reply:
I do yes. However it would seem that SEP-18 has already been accepted so it may not be worth transitioning the app namespace to an off-chain standard. My guess however is that we could probably move this all to a TOML storage solution without much if any issues.

Reply:
It's already been accepted but the only registered namespace right now is the `config` namespace which is specifically for use in SEPs, e.g. [SEP-29](https://github.com/stellar/stellar-protocol/blob/f3768980b558501e66cf9bd28c38bae248ae2252/ecosystem/sep-0029.md#specification).
https://github.com/stellar/stellar-protocol/blob/f3768980b558501e66cf9bd28c38bae248ae2252/ecosystem/sep-0018.md#L87-L92
This PR is the first introduction of `app`, `profile`, `wallet` into the registry. Technically they are mentioned in an example in the existing version of the SEP but that's just an example demonstrating how the SEP could be used and they are not in the registry at the bottom of the document.

Reply:
Quickly passing by before we stray away.

About the on-chain vs off-chain thing, naturally, this SEP is relevant _iff_ a distributed consensus is required for an application to function.

That's the case of `isMemoRequired` or signature coordination setup (SEP-19). You simply can't provide an equal level of robustness by using a TOML file.

For centralized applications, it goes without saying that off-chain storage must be preferred.

Reply:
Back to the main point, I'd agree on dropping `wallet` & `profile` until it gets specified in another document. I think we need to keep `app` though, as a way to keep vendors data contained away from the global namespace.

Reply:
My fear is this SEP will never gain adoption if it locks up funds in order to comply. I don't see a real reason to include this stuff on chain. `isMemoRequired` is a very specific use case and a response to how wallets are already used. This feels proactive in a way it doesn't need to be. Unless there's conversation and requests out there I'm unaware of. Would it be possible to get some wallet makers to chip in here?

Reply:
@tyvdh You can check [sep19](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0019.md) & [sep21](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0021.md) for other example of this.

The question of whether or not we need data on-chain has been much debated prior to the acceptation on this SEP, which got accepted since `isMemoRequired` is based on it.

This SEP is meant as a way to prevent namespace conflict. It doesn't mean that every user out there is going to have dozens of data entries defined.

Reply:
I think the reason to keep some data on-chain under the app namespace would be if a single service wanted accounts to register and store some information on their account, but if a single website/product is doing this they could also use SEP-10 to authenticate the account and then store the information off chain with the user authenticated.

The main value is where there is some common field that will be implemented with by many products at which point that interface is likely to become a SEP, and at that point the value could live under `config` like SEP-29.

As long as centralized apps keep their values outside the config namespace they are unlikely to clash with SEPs.

Reply:
The premise is that the global namespace (`config`, `wallet`, `profile`...`) is entirely reserved for SEPs. Not everything standard is meant to go under `config`.

This is similar to many computing system where the global namespace is a shared space that is meant to be kept unpolluted, while each modules gets its own dedicated space.

This is because vendor systems need to be compartmentalized anyway as else they would risk to conflict between each other.

Reply:
> I think the reason to keep some data on-chain under the app namespace would be if a single service wanted accounts to register and store some information on their account, but if a single website/product is doing this they could also use SEP-10 to authenticate the account and then store the information off chain with the user authenticated.

This is not the same: data stored on-chain can be mutated only with account signers consensus. That's something off-chain storage can't provide. on-chain storage retains the _trustless_ quality distributed ledgers offers, while off-chain storage introduces a single point of failure.

You need that on-chain storage for specific use case that are related to decentralized applications or account integrity.

For exemple, when it comes to multisig coordination, storing the setup on-chain is a matter of account integrity: it makes sense that signers consensus is required to required the signature coordination setup.

Reply:
Hmm. I think describing it as a single point of change is better than single point of failure and your flow is only true in a multisig scenario as the data can be changed just as easily if not more so when on chain than in a toml file if a wallet only has one signer. 

I think I can be on board for the app domain specs but wallet and profile seem preemptive and I'm not convinced they're better on-chain. We have a spec for storing meta data in TOML files and it works great and is widely used, I don't think your concerns are founded unless there are others who can chime in to support the need for `profile` and `wallet` metadata being stored on-chain.

Who else is asking for this change? Do they also feel strongly that account metadata should be stored on-chain?

There's also a difference between signing for a change and pushing data to the ledger. You could store signed data off-chain which couldn't be changed without a valid signature from the account. Just because something isn't on chain doesn't mean you can't be sure an account signed for it. So there are alternatives to on-chain which accomplish alleviating your stated concerns.

I don't want to beat a dead horse but I don't think it's unreasonable to query for alternatives and seek out other voices who you envision implementing and making use of this feature to chime in so we don't push standards in isolation which ultimately would not be implemented.

Reply:
Ultimately my vote is to simply include some clarifying language around cases where new namespaces may or should be created in conjunction with other SEPs. As other SEPs are made and wish to make use of on chain storage they should consider using the namespace which makes the most sense. So rather than creating an official list here preemptively we create the list as SEPs require it by looking at SEP-18 and realizing there can/should be other name spaces besides just `config.`

Reply:
Ok then, I edited the document accordingly.

Reply:
Is there a SEP using `app` yet? @leighmcculloch would you be using `app` for the work you're doing? Curious if it makes sense to have that included in this PR vs just update this SEP-18 as other SEPs are or should actually be using name spaces besides `config`. A registry is a great idea once names are in use but proactively I'm not sure is the right move. Curious though as I definitely see everyone potentially just force fitting `config` to their use case if more aren't at least suggested first. Thoughts?

Reply:
`app` is not meant to be used by any SEP. That's specifically meant for applications that can't/won't/don't want to go through the standardization process.

For example, a service wants to trustlessly record accounts KYC sha256sum for interoperability between applications relying on the said service. Standardizing into a SEP may not be a priority for them - as their initial goal is to get that MVP out.

Then they can put that data under `app.awesomekyc_com.id`.

> Curious though as I definitely see everyone potentially just force fitting `config` to their use case if more aren't at least suggested first. Thoughts?

I think we can simply enter the discussion whenever that happens.

Reply:
I'm not adding any data entries to any SEPs at the moment, but if we were adding them in a SEP we'd put them in config. I think @MisterTicot `app` is only useful in centralized use cases. Maybe having the namespace will just provide some organization so that folks who do want to use data entries for their application data will put it in one place that is clearly out of the way or anything else. I don't feel strongly about adding `app` but I'm not opposed to it. I agree we should defer `profile` and `wallet` for now.